### ISSUE TITLE
Community/Boards Engines Redesign

### ISSUE THREAD

**Issue by balmas** [sosol/sosol#112]
This issue describes a design proposal to modularize the communities and boards functionality into separate rails engines to facilitate their customization.

I'd like to propose a design which creates 2 new engines: a CommunitesEngine and a BoardsEngine.

The main SoSOL app would be responsible for wiring up the Community model classes provided by the CommunitesEngine with members and boards, and for specifying the class(es) used for each.  

Neither the Communities nor the Boards engine would have access to the Repository model class or be able to do git operations. All interactions with the git repository would be handled by the Publication class and controller.  The Communities engine would expose methods to add and remove boards, add and remove publications, add and remove members and "promote" a publication out of the community.  (Additional methods would be added for some of the new customizable functionality).  

As part of this, I'd like to implement a conceptual change in the main SoSOL app for publications. I'd like to abstract the current non-community workflow into a new Master community workflow, essentially turning the default case into a scenario where all the users of the application are members of a base master community, which has some special behavior, namely that it is the final stop in the workflow. After a publication makes it through this type of community's boards it can only go on to be committed to the master repo, with the original submitter attributed as the committer.  All publications would then always belong to a community, and community ownership would be assigned at creation time, with the default 'master' community being assigned by default if none is selected by the user (or creating agent) upon creation of a publication.

In addition, based upon the various use cases we've encountered for Perseids, I have identified 3 main base types of Communities that I'd like to be able to support:

- Master
As described above, this mimics the current "no community" workflow. By making this a type, we would now have the additional ability for more than one community to be able to promote publications on to the master repo without going through an end user

- EndUser
This offers the same functionality as the current communities workflow, where an end user must be supplied and publications are placed in this users workspace for editing after the community workflow completes.

- PassThrough
This is a new type of community which can be configured to pass publications on to another community and/or an external agent end point upon completion (this would improve the current Perseids/EAGLE workflow and be used for the proposed Perseids/Syriaca integration, as well as allow for communities to be chained together).

The main consideration driving me to make these named types, rather than just making all community behavior configurable attributes, is to enable the main SoSOL application to limit which community types it wants to allow in a given deployment, and to be able to properly ensure security controls on "master" communities.

Within these types, there would be a number of configurable attributes, including

- whether it is "open" or "closed" (where an "open" community is one that anyone can submit to, and a "closed" community is one that a member list must be configured by an admin)
- default result actions for the community's boards
- default email templates for the community's boards

You can see a rough sketch of a class diagram for the proposed CommunitiesEngine models and controllers at http://www.gliffy.com/go/publish/image/8909863/L.png. 

The rationale for making Boards a separate engine, rather than part of the Communities Engine, is that there are number of other customizations I would like to make possible for Boards (including allowing for the possibility of Boards to edit during the voting step, allowing for an assigner role, making reviewers of publications "sticky" for future submissions", etc.) and I think we want to be able to configure these customizations separately from the communities which use them.

My plan is to tackle this in stages, separating out communities first, then boards, and so I'll flesh out the boards design in a bit more detail once I'm further along with communities.

@ryanfb and @hcayless before I proceed though I'd like to get your feedback on the above, and know whether or not this seems like a reasonable way to go.  Thanks!

**Comment by balmas** [comment]
Just pinging @ryanfb and @hcayless to see if you had any thoughts before I start working on this ....

**Comment by ryanfb** [comment]
Hi, sorry for taking a bit to get back to you on this. I think the existing community/workflow model is a bit of a kludge, which makes things difficult to think around. One sticking point for me is "Neither the Communities nor the Boards engine would have access to the Repository model class or be able to do git operations" - maybe I'm missing something but I'm not sure how this would work (mostly for Boards). Boards don't really "do" any git operations directly themselves right now, but they do act as a (polymorphic) Publication owner type to hold publications and have a git repository associated with them, where the Publication class will frequently do something like `self.owner.git_operation` to perform a git operation in the owner repo (where `owner` is either a Board or User). I think removing that would require some major refactoring, so maybe that's not what you meant?

**Comment by balmas** [comment]
Hi - this is a good point, and something I misstated above. I believe the SoSOL app would have to configure the Boards model objects provided by the Boards engine with an instance of the Repository model class. So that all git operations would be handled through the Repository model associated with a Board, by the publicly available methods of the Repository class. I might have missed some cases here, but as far as I could see, the Board class never calls any git operations itself directly, they are all done through through the repository object. ( E.g. we have `publication.owner.repository.[git_operation]` not `publication.owner.[git_operations]`). Does this make more sense?

**Comment by ryanfb** [comment]
Yes, that makes sense.

I think my other major question is about whether we want to separate these into separate engines, or refactor in the "core" SoSOL app (and ideally, separate that into an engine). The idea behind the (long-promised) SoSOL engine separation (#43) was to hive off the "core" SoSOL classes (Identifier, Publication, Repository, Board, User, Community, etc.) into an engine so that different projects could include it and primarily modify it for their needs by subclassing these superclasses (so for example papyri.info would have a project which included the SoSOL engine and subclassed Identifier to DDBIdentifier, HGVIdentifier, etc., while Perseids would have a project which included the SoSOL engine and subclassed Identifier to CTSIdentifier etc.). So if #43 were done with the refactoring under discussion here as part of the SoSOL "core", project-specific customizations to Boards could be done by using that single SoSOL engine (maybe using [decorators](http://edgeguides.rubyonrails.org/engines.html#overriding-models-and-controllers) on the Board class instead of subclassing). Which seems to me like a more reasonable place to make the cut, instead of eventually having 3-4 repositories/engines (Boards engine, Communities engine, SoSOL "core" engine, and the final papyri.info/Perseids Rails projects which use these). Does that make sense, or is there another advantage to structuring it this way?

**Comment by balmas** [comment]
This is a good question.  I would be quite happy to do this refactoring in the core sosol app as this would definitely mean less work upfront for me right now, but I was also thinking about ways to lessen impact on papyri.info, etc. as requirements around communities and boards diverge.  I.e. much of what is driving this redesign from the Perseids side is the need to add in a bunch of new functionality around boards and communities sooner rather than later. Separating it out into an engine seems to me one way to reduce the divergence of the Perseids and master sosol branches. I was thinking to approach it so that first I would implement the communities and boards engines with the current base functionality from the master branch, and get master and the perseids branches switched over to using them, and then move on to the additional work needed for Perseids enhancements in functionality in a fork of just the engines.

**Comment by balmas** [comment]
For reference, here are the wishlist items driving this from the Perseids side:

https://github.com/perseids-project/perseids_docs/issues?q=is%3Aopen+is%3Aissue+milestone%3A%22Review+Workflow+Rewrite%22

some of these are minor but the big ones driving the design are:

- ability optionally enable "edit" while voting
- ability to link a specific publication with a specific reviewer 
- ability for community publications to go direct to master 
- ability to "template" communities and boards
- more customization options for board notifications

**Comment by ryanfb** [comment]
I think some/most of these are fine on the papyri.info side as well (just might go unused). The only one I see that might be tricky would be 'ability optionally enable "edit" while voting', which we may want some way to disable.

**Comment by balmas** [comment]
yes, that one (edit while voting) is definitely going to be tricky and I'm not positive how I'm going to do it yet :)

So, it sounds like you would rather then that I just proceed with refactoring the way communities work inside the core sosol app for now?

**Comment by ryanfb** [comment]
Yes - I think ultimately it will lead to a (hopefully) less-confusing codebase.

**Comment by balmas** [comment]
okey dokey :)  Will start this as a branch off master to try to keep it as clean as possible.

**Comment by balmas** [comment]
For reference - general target class diagram for new communities
http://www.gliffy.com/go/publish/image/8909863/L.png

**Comment by balmas** [comment]
Submit and Finalize workflows documented here: http://www.gliffy.com/go/publish/image/8872175/L.png

**Comment by balmas** [comment]
see issue #73 which needs to be addressed or the call to community.promote can be called when we don't want it to.

**Comment by balmas** [comment]
The second round of changes for this, to implement pass through communities, can be found in this branch: https://github.com/sosol/sosol/tree/passthroughcommunity
